content/blog/2025/11/advanced-porting-libraries-as-components/index.md

```diff
@@ -0,0 +1,621 @@
+
+**When to stub vs implement:**
+
+- **Stub**: When the function is called but its behavior isn't critical, or when a minimal implementation is sufficient
+- **Implement**: When the function is essential for correct operation
+
+#### Function Namespacing
+
+When building on Linux for testing, you may need to namespace functions to avoid conflicts with system libraries. The sock_utils component does this:
+
+```c
+#ifdef CONFIG_IDF_TARGET_LINUX
+// namespace with esp_ on linux to avoid conflict of symbols
+#define socketpair esp_socketpair
+#define pipe esp_pipe
+#endif
+```
+
+This ensures that when building on Linux, the ESP-IDF implementations use prefixed names, avoiding conflicts with the system's POSIX implementations.
+
+## Step 5: Fine-tune
+
+### Managing Compiler Warnings
+
+#### Component-Level Suppression
+
+As covered in the basic porting article, you can suppress compiler warnings for an entire component using `target_compile_options()`:
+
+```cmake
+target_compile_options(${COMPONENT_LIB} PRIVATE -Wno-char-subscripts)
+```
+
+This approach works well when the warning applies broadly across the component, but it can be too broad when only specific files trigger warnings.
+
+#### File-Level Suppression
+
+For more granular control, ESP-IDF allows you to suppress warnings on a per-file basis using `set_source_files_properties()`. This is particularly useful when:
+
+- Only a few files in a large library trigger warnings
+- You want to keep strict warnings enabled for the rest of the component
+- Different files need different warning suppressions
+
+The mosquitto library provides a good example of this technique. Some mosquitto source files unconditionally define `_GNU_SOURCE`, which collides with the ESP-IDF build system and produces redefinition warnings. Instead of suppressing this warning for the entire component, we can target only the offending files:
+
+```cmake
+# Some mosquitto source unconditionally define `_GNU_SOURCE` which collides with IDF build system
+# producing warning: "_GNU_SOURCE" redefined
+# This workarounds this issue by undefining the macro for the selected files
+set(sources_that_define_gnu_source ${m_src_dir}/loop.c ${m_src_dir}/mux_poll.c)
+foreach(offending_src ${sources_that_define_gnu_source})
+    set_source_files_properties(${offending_src} PROPERTIES COMPILE_OPTIONS "-U_GNU_SOURCE")
+endforeach()
+```
+
+You can also use this technique to suppress specific warnings for individual files:
+
+```cmake
+# Suppress format warnings for specific files
+set_source_files_properties(${m_src_dir}/logging.c PROPERTIES COMPILE_OPTIONS "-Wno-format")
+```
+
+Or combine multiple options:
+
+```cmake
+set_source_files_properties(${m_src_dir}/file.c PROPERTIES
+    COMPILE_OPTIONS "-Wno-format;-Wno-unused-variable")
+```
+
+**When to use file-level vs component-level:**
+
+- **Component-level**: Use when the warning pattern is consistent across most or all files, or when the library is small
```

Comment:
When you have a bullet point with a sentence, please add a period `.` on each item. Please look over the other bullet points in your article to add the period when needed.